src/simulator/main_utils.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+#include "../common/AlgorithmRegistrar.h"
+#include "my_simulator.h"
+#include "robot.h"
+#include <dlfcn.h>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+
+struct SimInfo {
+    MySimulator simulator;
+    std::string house_file_name;
+    std::string house_output_path;
+};
+
+struct AlgoInfo {
+    std::unique_ptr<AbstractAlgorithm> algorithm;
+    std::string algo_file_path;
+    void *handle; // Save the handle so we can dlclose it later.
+    std::string name;
+};
+
+struct SummaryInfo {
+    std::string house_file_name;
+    std::string algo_name;
+    std::size_t score;
+};
+
+namespace fs = std::filesystem;
+
+void processHouseFile(const fs::path &houseFile,
+                      std::vector<SimInfo> &simulators) {
+
+    std::ifstream file(houseFile);
+    fs::path errorFilePath =
+        fs::current_path() / (houseFile.stem().string() + ".error");
+
+    if (!file.is_open()) {
+        std::ofstream errorFile(errorFilePath);
+        errorFile << "Error: Could not open house file: " << houseFile
+                  << std::endl;
+    }
+    try {
+        SimInfo simInfo;
+        simInfo.simulator.readHouseFile(houseFile.string());
+        simInfo.house_file_name = houseFile.stem().string();
+        std::string outputFile = addPrefixToFileName(houseFile.stem().string());
+        simInfo.house_output_path = "../../../output/" + outputFile;
+        simulators.push_back(std::move(simInfo));
+
+    } catch (const std::exception &e) {
+        std::ofstream errorFile(errorFilePath);
+        errorFile << "Error: " << e.what() << std::endl;
+    }
+}
+
+void processAlgorithmFile(const fs::path &algoFile,
+                          std::vector<AlgoInfo> &algorithms) {
+
+    std::size_t curr_count =
+        AlgorithmRegistrar::getAlgorithmRegistrar().count();
+    void *handle = dlopen(algoFile.c_str(), RTLD_LAZY);
+    if ((curr_count !=
+         AlgorithmRegistrar::getAlgorithmRegistrar().count() - 1) ||
+        !handle) {
+        fs::path errorFilePath =
+            fs::current_path() / (algoFile.stem().string() + ".error");
+        std::ofstream errorFile(errorFilePath);
+        errorFile << "Error: Could not load algorithm file: " << algoFile
+                  << "\n"
+                  << dlerror() << std::endl;
+        return;
+    }
+
+    AlgoInfo algoInfo;
+    algoInfo.algorithm = std::move(
+        (*(AlgorithmRegistrar::getAlgorithmRegistrar().end() - 1)).create());
+    algoInfo.handle = handle;
+    algoInfo.algo_file_path = algoFile.string();
+    algoInfo.name =
+        (*(AlgorithmRegistrar::getAlgorithmRegistrar().end() - 1)).name();
+    algorithms.push_back(std::move(algoInfo));
+}
+void processFilesInDirectory(const fs::path &dirPath,
```

Comment:
its weird that the function does 2 different things,and only uses one of the argument vectors at a time.
have the smaller function output the struct instead of pushing to the given vector,

and this one initialize and output the vector so the caller doesnt have to(main).

we can even use this temple:

template <typename Iterator, typename Func>
std::vector<typename std::result_of<Func(typename std::iterator_traits<Iterator>::value_type)>::type>
createVectorFromIterator(Iterator begin, Iterator end, Func mapFunc) {
    using ValueType = typename std::result_of<Func(typename std::iterator_traits<Iterator>::value_type)>::type;
    std::vector<ValueType> result;
    result.reserve(std::distance(begin, end));

    for (Iterator it = begin; it != end; ++it) {
        result.push_back(mapFunc(*it));
    }

    return result;
}
and then you can:


 std::vector<int> myVector = createVectorFromIterator(dir_iterator.begin(), dir_iterator.end(), processFile);

feel free to do things differently, but hide as much as possible from main, and preferr doing allocation inside functions rather then functions that take destination as an argument (ugly and redundent)

Reply:
Updated, check new commit